py: Remove calls to file reader functions when these are disabled

sburton84 opened the pull request:
If `MICROPY_PERSISTENT_CODE_LOAD` or `MICROPY_ENABLE_COMPILER` are enabled then code gets enabled that calls file reading functions which are disabled if neither `MICROPY_READER_POSIX` or `MICROPY_READER_VFS` is set. These calls should also be disabled if neither of these is set.

Specifically, this happens in 3 instances:

1. In `persistentcode.c` if `MICROPY_PERSISTENT_CODE_LOAD` is set then the function `mp_raw_code_load_file` gets defined regardless of whether `MICROPY_READER_POSIX` or `MICROPY_READER_VFS` is set, but then tries to call `mp_reader_new_file` which will only be defined if one of these is set.
2. In `builtinimport.c`, if `MICROPY_PERSISTENT_CODE_LOAD` is set, the `do_load` function tries to call the `mp_raw_code_load_file` mentioned above.
3. In `builtinimport.c` if `MICROPY_ENABLE_COMPILER` is set the `do_load` function tries to call the `mp_lexer_new_from_file` function without checking whether `MICROPY_READER_POSIX` or `MICROPY_READER_VFS` is set. But in `lexer.c` the  function is disabled if neither of these is set.

This change simply adds extra `#if MICROPY_READER_POSIX || MICROPY_READER_VFS` checks around these sections to disable them when necessary.

Commit:
py: Remove calls to file reader functions when these are disabled

If MICROPY_PERSISTENT_CODE_LOAD or MICROPY_ENABLE_COMPILER are enabled
then code gets enabled that calls file reading functions which may be
disabled if no readers have been implemented.

To fix this, introduce a MICROPY_HAS_READER variable, which is
automatically set if MICROPY_READER_POSIX or MICROPY_READER_VFS is set
but can also be manually set if a custom reader is being implemented.
Then disable the file reading calls if this is set.

pfalcon commented:
That's a questionable change. MICROPY_READER_POSIX and MICROPY_READER_VFS merely enable one of pre-provided readers. A user may define their own reader implementation and want to use it.

stinos commented:
Agreed, unless this gets done as a configuration option in mpconfigport.h like
```
#ifndef MICROPY_HAS_READERS
#define MICROPY_HAS_READERS (MICROPY_READER_POSIX || MICROPY_READER_VFS)
#endif
```
(or something like that, just thinking aloud here), it's too limiting.

pfalcon commented:
> There is already an identical check in lexer.c around the mp_lexer_new_from_file function, so if this change is questionable then that check surely is as well.

Quite possible.

> I think adding a MICROPY_HAS_READERS setting like described above makes sense.

+1. Should be MICROPY_HAS_READER (singular) I guess.

sburton84 commented:
I've added the MICROPY_HAS_READER variable and changed the checks to use this, including the one in `lexer.c`.

Also had to extend a couple of the other checks in `builtinimport.c` to prevent unused function and unused variable errors in the CI. The check around `file_str` is getting a bit unwieldy, is this okay as it is?

Just looking at the remaining CI errors, one of them seems to be that the code size has increased by 32 bytes, I'm not sure why that would be as I'm mostly just disabling bits?

sburton84 commented:
The change in code size was caused by the `mp_raise_msg` call being moved outside the else branch of the `#if MICROPY_ENABLE_COMPILER`. I've moved this back into the else and all the checks seem to be passing now.

dpgeorge commented:
Thanks for the contribution, it looks pretty good.

I think the config option should be `MICROPY_HAS_FILE_READER` because what it's really doing is telling whether the `mp_reader_new_file()` function is available or not.  Even when it's not available the reader is still there, eg via `mp_reader_new_mem()`.

One question: in `do_load()` it looks like it's now possible that this function just returns without doing anything, because there's no longer an unconditional piece of code at the end of it.  Can you see what are the conditions in which this function would just return silently without attempting to load anything?

> The check around `file_str` is getting a bit unwieldy, is this okay as it is?

Yes it's ok.

sburton84 commented:
> I think the config option should be `MICROPY_HAS_FILE_READER` because what it's really doing is telling whether the `mp_reader_new_file()` function is available or not. Even when it's not available the reader is still there, eg via `mp_reader_new_mem()`.

I've renamed this variable accordingly.

> One question: in `do_load()` it looks like it's now possible that this function just returns without doing anything, because there's no longer an unconditional piece of code at the end of it. Can you see what are the conditions in which this function would just return silently without attempting to load anything?

Yeah, originally I moved the `mp_raise_msg` message right to the end outside of the `#else` which is the simplest solution but also increased the code size unnecessarily. I've changed it now so the `MICROPY_HAS_FILE_READER` check is in each of the `MICROPY_PERSISTENT_CODE_LOAD` and `MICROPY_ENABLE_COMPILER` checks so the `mp_raise_msg` has been moved back into the `#else` of the latter, I think this should be correct?

sburton84 commented:
Just realised this probably isn't quite right still. It looks like a port implementing their own reader will implement their own `mp_lexer_new_from_file` function, but if `MICROPY_HAS_FILE_READER` is set, which they would want to set to enable the parts the use this, then the `mp_lexer_new_from_file` function in `lexer.c` will get enabled. I guess this is why the check in `lexer.c` was `MICROPY_READER_POSIX || MICROPY_READER_VFS` and I think I should change it back to that.

Also I suspect the 32-byte size increase is because of the `"script compilation not supported"` string, which is 32 characters long. The ports don't have `MICROPY_HAS_FILE_READER` set, but implement their own `mp_lexer_new_from_file` which just returns ENOENT. It might make sense for me to remove these `mp_lexer_new_from_file` implementations if it will now raise the `"script compilation not supported"` error instead of trying to call them? But I'm not sure how this change would interact with the `mp_lexer_new_from_file` implementation in `lib/memzip/lexermemzip.c`? I guess if you were using that you wouldn't want the call to it to be disabled if `MICROPY_HAS_FILE_READER` is not set?

sburton84 commented:
I think it might finally be correct now. I reverted the change in `lexer.c` so it still checks `MICROPY_READER_POSIX || MICROPY_READER_VFS` before defining `mp_lexer_new_from_file`, then a port implementing their own reader can implement their own `mp_lexer_new_from_file`. I also removed the `MICROPY_HAS_FILE_READER` check on the `MICROPY_ENABLE_COMPILER` part of `do_load` so the `mp_lexer_new_from_file` call remains but calls into the versions which raise `ENOENT` in ports which don't support file loading.

dpgeorge commented:
Thanks for looking deeper into this, the change looks a lot cleaner now and I assume the change fixes the issues originally reported.

sburton84 commented:
> I assume the change fixes the issues originally reported.

Yeah, these changes fix issues 1 and 2 from above by preventing it from trying to call `mp_reader_new_file` if no reader exists. And issue 3 was fixed by instead just separately providing my own `mp_lexer_new_from_file` implementation.

dpgeorge commented:
Merged in e33bc59712aa484107dc1cd4009dd8ab2da6dcb2
